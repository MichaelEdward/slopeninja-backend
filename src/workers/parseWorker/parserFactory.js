import cheerio from 'cheerio';

export const removeBackSlashes = (text) => {
  const decoded = text ? text.toString().replace(/\\/g, '') : null;
  return decoded;
}

export const createHtmlParser = (key, parser, preParser = t => t) => async (html) => {
  const $ = cheerio.load(preParser(html));

  const keyData = await parser($);

  return {
    [key]: keyData
  };
}

export const createJSONParser = (key, parser) => async (data = "{}") => {
  let json = JSON.parse(data);

  let d;
  if (json) {
    d = json;
  }

  const keyData = await parser(d);

  return {
    [key]: keyData
  };
}
