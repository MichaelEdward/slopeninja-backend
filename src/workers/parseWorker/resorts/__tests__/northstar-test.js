import fs from 'fs';
import {
  parseNorthstarSnow,
  parseNorthstarLifts,
  parseNorthstarLiftList,
  parseNorthstarTrailList,
  parseNorthstarTrails
} from '../northstar';
import { createHtmlParser, createJSONParser } from '../../parserFactory';

test('fetches Northstar snow data correctly', async () => {
  const htmlText = fs.readFileSync(`${__dirname}/fixtures/northstar-weather.html`);

  const resortData = await createHtmlParser('snow', parseNorthstarSnow)(htmlText);
  expect(resortData).toEqual({
    snow: {
      status: 'closed',
      weatherIcon: null,
      temprature: null,
      baseCondition: null,
      newSnow: 2,
      snowDepthBase: null,
      snowDepthSummit: 0,
    },
  });
})

test('fetches all null for nonexisting snow values', async () => {
  const resortData = await createHtmlParser('snow', parseNorthstarSnow)('<html></html>');
  expect(resortData).toEqual({
    snow: {
      status: null,
      weatherIcon: null,
      temprature: null,
      baseCondition: null,
      newSnow: null,
      snowDepthBase: null,
      snowDepthSummit: null,
    }
  });
});

test('fetches Northstar lifts data correctly', async () => {
  const htmlText = fs.readFileSync(`${__dirname}/fixtures/northstar-lifts.html`);
  const resortData = await createHtmlParser('lifts', parseNorthstarLifts)(htmlText);
  expect(resortData).toEqual({
    lifts: {
      total: 20,
      open: 0,
    }
  });
});

test('fetches all null for nonexisting lift values', async () => {
  const resortData = await createHtmlParser('lifts', parseNorthstarLifts)('<html></html>');
  expect(resortData).toEqual({
    lifts: {
      total: null,
      open: null,
    }
  });
});

test('fetches Northstar trails data correctly', async () => {
  const htmlText = fs.readFileSync(`${__dirname}/fixtures/northstar-lifts.html`);
  const resortData = await createHtmlParser('trails', parseNorthstarTrails)(htmlText);
  expect(resortData).toEqual({
    trails: {
      total: 100,
      open: 0,
    }
  });
});

test('fetches all null for nonexisting trails values', async () => {
  const resortData = await createHtmlParser('trails', parseNorthstarTrails)('<html></html>');
  expect(resortData).toEqual({
    trails: {
      total: null,
      open: null,
    }
  });
});

test('fetches Northstar lift list correctly', async () => {
  const htmlText = fs.readFileSync(`${__dirname}/fixtures/northstar-lifts.html`);
  const resortData = await createHtmlParser('liftList', parseNorthstarLiftList)(htmlText);
  expect(resortData).toMatchSnapshot();
});

test('fetches all null for nonexisting lift list values', async () => {
  const resortData = await createHtmlParser('liftList', parseNorthstarLiftList)('<html></html>');
  expect(resortData).toMatchObject({ liftList: [] });
});

test('fetches Northstar trail list correctly', async () => {
  const htmlText = fs.readFileSync(`${__dirname}/fixtures/northstar-lifts.html`);
  const resortData = await createHtmlParser('trailList', parseNorthstarTrailList)(htmlText);
  expect(resortData).toMatchSnapshot();
});

test('fetches all null for nonexisting lift list values', async () => {
  const resortData = await createHtmlParser('trailList', parseNorthstarTrailList)('<html></html>');
  expect(resortData).toMatchObject({ trailList: [] });
});
