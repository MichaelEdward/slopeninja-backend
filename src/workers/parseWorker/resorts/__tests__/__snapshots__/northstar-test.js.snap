// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`fetches Northstar lift list correctly 1`] = `
Object {
  "liftList": Array [
    Object {
      "category": null,
      "name": "Arrow Express",
      "status": "closed",
    },
    Object {
      "category": null,
      "name": "Backside Express",
      "status": "closed",
    },
    Object {
      "category": null,
      "name": "Big Dipper Carpet",
      "status": "closed",
    },
    Object {
      "category": null,
      "name": "Big Springs Express Gondola",
      "status": "closed",
    },
    Object {
      "category": null,
      "name": "Comstock Express",
      "status": "closed",
    },
    Object {
      "category": null,
      "name": "Highlands Gondola",
      "status": "closed",
    },
    Object {
      "category": null,
      "name": "Little Dipper Carpet",
      "status": "closed",
    },
    Object {
      "category": null,
      "name": "Lookout Link Platter",
      "status": "closed",
    },
    Object {
      "category": null,
      "name": "Martis Camp Express",
      "status": "closed",
    },
    Object {
      "category": null,
      "name": "Orion's Belt",
      "status": "closed",
    },
    Object {
      "category": null,
      "name": "Pegasus Carpet",
      "status": "closed",
    },
    Object {
      "category": null,
      "name": "Promised Land Express",
      "status": "closed",
    },
    Object {
      "category": null,
      "name": "Rendezvous Triple",
      "status": "closed",
    },
    Object {
      "category": null,
      "name": "Tahoe Zephyr Express",
      "status": "closed",
    },
    Object {
      "category": null,
      "name": "The Big Easy Quad",
      "status": "closed",
    },
    Object {
      "category": null,
      "name": "Timberline Triple",
      "status": "closed",
    },
    Object {
      "category": null,
      "name": "Tubing Tow",
      "status": "closed",
    },
    Object {
      "category": null,
      "name": "Ursa Major Carpet",
      "status": "closed",
    },
    Object {
      "category": null,
      "name": "Village Express",
      "status": "closed",
    },
    Object {
      "category": null,
      "name": "Vista Express",
      "status": "closed",
    },
  ],
}
`;

exports[`fetches Northstar trail list correctly 1`] = `
Object {
  "trailList": Array [
    Object {
      "category": null,
      "level": "intermediate",
      "name": "Lookout Bypass",
      "status": "closed",
    },
    Object {
      "category": null,
      "level": "intermediate",
      "name": "Schwarzstrasse",
      "status": "closed",
    },
    Object {
      "category": null,
      "level": "intermediate",
      "name": "Washoe",
      "status": "closed",
    },
    Object {
      "category": null,
      "level": "advanced",
      "name": "Boca",
      "status": "closed",
    },
    Object {
      "category": null,
      "level": "advanced",
      "name": "Gooseneck",
      "status": "closed",
    },
    Object {
      "category": null,
      "level": "advanced",
      "name": "Lookout Glade",
      "status": "closed",
    },
    Object {
      "category": null,
      "level": "advanced",
      "name": "Martis",
      "status": "closed",
    },
    Object {
      "category": null,
      "level": "advanced",
      "name": "Prosser",
      "status": "closed",
    },
    Object {
      "category": null,
      "level": "advanced",
      "name": "Stampede",
      "status": "closed",
    },
    Object {
      "category": null,
      "level": "advanced",
      "name": "Sugar Pine Glade",
      "status": "closed",
    },
    Object {
      "category": null,
      "level": "advanced",
      "name": "The Camp Glade",
      "status": "closed",
    },
    Object {
      "category": null,
      "level": "advanced",
      "name": "Upper Martis",
      "status": "closed",
    },
    Object {
      "category": null,
      "level": "begineer",
      "name": "Lower Main Street",
      "status": "closed",
    },
    Object {
      "category": null,
      "level": "begineer",
      "name": "Lumberjack",
      "status": "closed",
    },
    Object {
      "category": null,
      "level": "begineer",
      "name": "Skid Trail",
      "status": "closed",
    },
    Object {
      "category": null,
      "level": "intermediate",
      "name": "Ax Handle",
      "status": "closed",
    },
    Object {
      "category": null,
      "level": "intermediate",
      "name": "Cascades",
      "status": "closed",
    },
    Object {
      "category": null,
      "level": "intermediate",
      "name": "Cat's Face",
      "status": "closed",
    },
    Object {
      "category": null,
      "level": "intermediate",
      "name": "Drop Off",
      "status": "closed",
    },
    Object {
      "category": null,
      "level": "intermediate",
      "name": "East Ridge",
      "status": "closed",
    },
    Object {
      "category": null,
      "level": "intermediate",
      "name": "Goldmine",
      "status": "closed",
    },
    Object {
      "category": null,
      "level": "intermediate",
      "name": "Logger's Loop",
      "status": "closed",
    },
    Object {
      "category": null,
      "level": "intermediate",
      "name": "Lookout Road",
      "status": "closed",
    },
    Object {
      "category": null,
      "level": "intermediate",
      "name": "Luggis",
      "status": "closed",
    },
    Object {
      "category": null,
      "level": "intermediate",
      "name": "Stump Alley",
      "status": "closed",
    },
    Object {
      "category": null,
      "level": "intermediate",
      "name": "The Flume",
      "status": "closed",
    },
    Object {
      "category": null,
      "level": "intermediate",
      "name": "Upper Jibboom",
      "status": "closed",
    },
    Object {
      "category": null,
      "level": "intermediate",
      "name": "Upper Main Street",
      "status": "closed",
    },
    Object {
      "category": null,
      "level": "intermediate",
      "name": "West Ridge",
      "status": "closed",
    },
    Object {
      "category": null,
      "level": "intermediate",
      "name": "Woodchuck",
      "status": "closed",
    },
    Object {
      "category": null,
      "level": "advanced",
      "name": "Comstock Glade",
      "status": "closed",
    },
    Object {
      "category": null,
      "level": "advanced",
      "name": "Corridor",
      "status": "closed",
    },
    Object {
      "category": null,
      "level": "advanced",
      "name": "Cowboy Pass",
      "status": "closed",
    },
    Object {
      "category": null,
      "level": "advanced",
      "name": "Crosscut",
      "status": "closed",
    },
    Object {
      "category": null,
      "level": "advanced",
      "name": "Delight",
      "status": "closed",
    },
    Object {
      "category": null,
      "level": "advanced",
      "name": "Dutchman",
      "status": "closed",
    },
    Object {
      "category": null,
      "level": "advanced",
      "name": "Flying Squirrel",
      "status": "closed",
    },
    Object {
      "category": null,
      "level": "advanced",
      "name": "Grouse Alley",
      "status": "closed",
    },
    Object {
      "category": null,
      "level": "advanced",
      "name": "Hornet's Nest",
      "status": "closed",
    },
    Object {
      "category": null,
      "level": "advanced",
      "name": "Lower Jibboom",
      "status": "closed",
    },
    Object {
      "category": null,
      "level": "advanced",
      "name": "Powder Bowl",
      "status": "closed",
    },
    Object {
      "category": null,
      "level": "advanced",
      "name": "Rendezvous Glade",
      "status": "closed",
    },
    Object {
      "category": null,
      "level": "advanced",
      "name": "Sawmill Glade",
      "status": "closed",
    },
    Object {
      "category": null,
      "level": "advanced",
      "name": "Spring Board",
      "status": "closed",
    },
    Object {
      "category": null,
      "level": "advanced",
      "name": "Surprise",
      "status": "closed",
    },
    Object {
      "category": null,
      "level": "advanced",
      "name": "The Chute",
      "status": "closed",
    },
    Object {
      "category": null,
      "level": "advanced",
      "name": "The Plunge",
      "status": "closed",
    },
    Object {
      "category": null,
      "level": "advanced",
      "name": "Tonini's",
      "status": "closed",
    },
    Object {
      "category": null,
      "level": "begineer",
      "name": "Bearly",
      "status": "closed",
    },
    Object {
      "category": null,
      "level": "begineer",
      "name": "Easy Street",
      "status": "closed",
    },
    Object {
      "category": null,
      "level": "begineer",
      "name": "Lower Lion's Way",
      "status": "closed",
    },
    Object {
      "category": null,
      "level": "begineer",
      "name": "The Gulch",
      "status": "closed",
    },
    Object {
      "category": null,
      "level": "begineer",
      "name": "Woodcutter",
      "status": "closed",
    },
    Object {
      "category": null,
      "level": "intermediate",
      "name": "Boondocks",
      "status": "closed",
    },
    Object {
      "category": null,
      "level": "intermediate",
      "name": "Christmas Tree",
      "status": "closed",
    },
    Object {
      "category": null,
      "level": "intermediate",
      "name": "Gateway",
      "status": "closed",
    },
    Object {
      "category": null,
      "level": "intermediate",
      "name": "Goldmine",
      "status": "closed",
    },
    Object {
      "category": null,
      "level": "intermediate",
      "name": "Home Run",
      "status": "closed",
    },
    Object {
      "category": null,
      "level": "intermediate",
      "name": "Hoot Owl",
      "status": "closed",
    },
    Object {
      "category": null,
      "level": "intermediate",
      "name": "Lookout Bypass",
      "status": "closed",
    },
    Object {
      "category": null,
      "level": "intermediate",
      "name": "Northern Lights",
      "status": "closed",
    },
    Object {
      "category": null,
      "level": "intermediate",
      "name": "Overland Trail",
      "status": "closed",
    },
    Object {
      "category": null,
      "level": "intermediate",
      "name": "Pioneer",
      "status": "closed",
    },
    Object {
      "category": null,
      "level": "intermediate",
      "name": "Sodergren's",
      "status": "closed",
    },
    Object {
      "category": null,
      "level": "intermediate",
      "name": "The Glades",
      "status": "closed",
    },
    Object {
      "category": null,
      "level": "intermediate",
      "name": "The Gully",
      "status": "closed",
    },
    Object {
      "category": null,
      "level": "intermediate",
      "name": "Timberline",
      "status": "closed",
    },
    Object {
      "category": null,
      "level": "intermediate",
      "name": "Upper Lion's Way",
      "status": "closed",
    },
    Object {
      "category": null,
      "level": "intermediate",
      "name": "Upper Pioneer",
      "status": "closed",
    },
    Object {
      "category": null,
      "level": "advanced",
      "name": "Cowboy Pass",
      "status": "closed",
    },
    Object {
      "category": null,
      "level": "advanced",
      "name": "The Face",
      "status": "closed",
    },
    Object {
      "category": null,
      "level": "begineer",
      "name": "Lil Stash",
      "status": "closed",
    },
    Object {
      "category": null,
      "level": "intermediate",
      "name": "Forerunner",
      "status": "closed",
    },
    Object {
      "category": null,
      "level": "intermediate",
      "name": "Moonshine",
      "status": "closed",
    },
    Object {
      "category": null,
      "level": "intermediate",
      "name": "Pinball",
      "status": "closed",
    },
    Object {
      "category": null,
      "level": "intermediate",
      "name": "Pipeline",
      "status": "closed",
    },
    Object {
      "category": null,
      "level": "intermediate",
      "name": "Playground",
      "status": "closed",
    },
    Object {
      "category": null,
      "level": "intermediate",
      "name": "Sidewinder",
      "status": "closed",
    },
    Object {
      "category": null,
      "level": "intermediate",
      "name": "Straits",
      "status": "closed",
    },
    Object {
      "category": null,
      "level": "intermediate",
      "name": "The Stash",
      "status": "closed",
    },
    Object {
      "category": null,
      "level": "intermediate",
      "name": "Castle Peak",
      "status": "closed",
    },
    Object {
      "category": null,
      "level": "intermediate",
      "name": "Drifter",
      "status": "closed",
    },
    Object {
      "category": null,
      "level": "intermediate",
      "name": "Drifter Connector",
      "status": "closed",
    },
    Object {
      "category": null,
      "level": "intermediate",
      "name": "The Islands",
      "status": "closed",
    },
    Object {
      "category": null,
      "level": "advanced",
      "name": "Back Door",
      "status": "closed",
    },
    Object {
      "category": null,
      "level": "advanced",
      "name": "Burnout",
      "status": "closed",
    },
    Object {
      "category": null,
      "level": "advanced",
      "name": "Challenger",
      "status": "closed",
    },
    Object {
      "category": null,
      "level": "advanced",
      "name": "Down Under",
      "status": "closed",
    },
    Object {
      "category": null,
      "level": "advanced",
      "name": "Follow Me",
      "status": "closed",
    },
    Object {
      "category": null,
      "level": "advanced",
      "name": "Iron Horse",
      "status": "closed",
    },
    Object {
      "category": null,
      "level": "advanced",
      "name": "Lower Burnout",
      "status": "closed",
    },
    Object {
      "category": null,
      "level": "advanced",
      "name": "Monument Glade",
      "status": "closed",
    },
    Object {
      "category": null,
      "level": "advanced",
      "name": "Polaris",
      "status": "closed",
    },
    Object {
      "category": null,
      "level": "advanced",
      "name": "Promised Land",
      "status": "closed",
    },
    Object {
      "category": null,
      "level": "advanced",
      "name": "Rail Splitter",
      "status": "closed",
    },
    Object {
      "category": null,
      "level": "advanced",
      "name": "Sierra Grande",
      "status": "closed",
    },
    Object {
      "category": null,
      "level": "advanced",
      "name": "The Rapids",
      "status": "closed",
    },
    Object {
      "category": null,
      "level": "advanced",
      "name": "Why Not",
      "status": "closed",
    },
    Object {
      "category": null,
      "level": "begineer",
      "name": "Condo Run",
      "status": "closed",
    },
    Object {
      "category": null,
      "level": "begineer",
      "name": "Village Run",
      "status": "closed",
    },
    Object {
      "category": null,
      "level": "intermediate",
      "name": "Coyote Fork",
      "status": "closed",
    },
    Object {
      "category": null,
      "level": "intermediate",
      "name": "Home Run",
      "status": "closed",
    },
    Object {
      "category": null,
      "level": "intermediate",
      "name": "Shortcut",
      "status": "closed",
    },
    Object {
      "category": null,
      "level": "intermediate",
      "name": "The Woods",
      "status": "closed",
    },
  ],
}
`;
